compiler/rustc_data_structures/src/sip128.rs

```diff
@@ -51,271 +61,378 @@ macro_rules! compress {
+// the integer write buffer processing path. The value that fills the buffer can
+// be written with a statically sized write that may spill over into the spill.
+// After the buffer is processed, the part of the value that spilled over can
+// written from the spill to the beginning of the buffer with another statically
+// sized write. Due to static sizes, this scheme performs better than copying
+// the exact number of bytes needed into the end and beginning of the buffer.
+//
+// The buffer is uninitialized, which improves performance, but may preclude
+// efficient implementation of alternative approaches. The improvement is not so
+// large that an alternative approach should be disregarded because it cannot be
+// efficiently implemented with an uninitialized buffer. On the other hand, an
+// uninitialized buffer may become more important should a larger one be used.
+//
+// # Platform Dependence
+//
+// The SipHash algorithm operates on byte sequences. It parses the input stream
+// as 8-byte little-endian integers. Therefore, given the same byte sequence, it
+// produces the same result on big- and little-endian hardware.
+//
+// However, the Hasher trait has methods which operate on multi-byte integers.
+// How they are converted into byte sequences can be endian-dependent (by using
+// native byte order) or independent (by consistently using either LE or BE byte
+// order). It can also be `isize` and `usize` size dependent (by using the
+// native size), or independent (by converting to a common size), supposing the
+// values can be represented in 32 bits.
+//
+// In order to make `SipHasher128` consistent with `SipHasher` in libstd, we
+// choose to do the integer to byte sequence conversion in the platform-
+// dependent way. Clients can achieve (nearly) platform-independent hashing by
+// widening `isize` and `usize` integers to 64 bits on 32-bit systems and
+// byte-swapping integers on big-endian systems before passing them to the
+// writing functions. This causes the input byte sequence to look identical on
+// big- and little- endian systems (supposing `isize` and `usize` values can be
+// represented in 32 bits), which ensures platform-independent results.
 impl SipHasher128 {
     #[inline]
     pub fn new_with_keys(key0: u64, key1: u64) -> SipHasher128 {
-        let mut state = SipHasher128 {
-            k0: key0,
-            k1: key1,
-            length: 0,
-            state: State { v0: 0, v1: 0, v2: 0, v3: 0 },
-            tail: 0,
-            ntail: 0,
+        let mut hasher = SipHasher128 {
+            nbuf: 0,
+            buf: MaybeUninit::uninit_array(),
+            state: State {
+                v0: key0 ^ 0x736f6d6570736575,
+                // The XOR with 0xee is only done on 128-bit algorithm version.
+                v1: key1 ^ (0x646f72616e646f6d ^ 0xee),
+                v2: key0 ^ 0x6c7967656e657261,
+                v3: key1 ^ 0x7465646279746573,
+            },
```

Comment:
I see that these constants and computations are copied from `reset`, but I don't understand why, when the old `new_with_keys` just used zeroes. Can you explain in a comment?

Reply:
The old `new_with_keys` initialized with zeroes, but then called `reset` to overwrite them with the XOR results. This just initializes them with the XOR results directly. I ditched `reset` because it was unused, and didn't seem likely to be used in the future. It's not likely to be used because `finish128` consumes self, so we won't reuse a hasher if it ever computes a hash.